File: rdvp/src/main/java/com/grdvp/service/DemandeRDVService.java
package com.grdvp.service;

import com.grdvp.entity.DemandeRDV;
import com.grdvp.entity.Patient;
import com.grdvp.entity.Specialite;
import com.grdvp.entity.Statut;
import com.grdvp.repository.DemandeRDVRepositoryImpl;
import com.grdvp.repository.PatientRepositoryImpl;
import com.grdvp.service.interfaces.DemandeRDVServiceImpl;

import java.time.LocalDateTime;
import java.util.List;

public class DemandeRDVService implements DemandeRDVServiceImpl {

    private static DemandeRDVService instance;
    private final DemandeRDVRepositoryImpl demandeRepo;
    private final PatientRepositoryImpl patientRepo;

    private DemandeRDVService() {
        this.demandeRepo = DemandeRDVRepositoryImpl.getInstance();
        this.patientRepo = PatientRepositoryImpl.getInstance();
    }

    public static DemandeRDVService getInstance() {
        if (instance == null) {
            instance = new DemandeRDVService();
        }
        return instance;
    }

    public DemandeRDV createDemandeRDV(Patient patient, Specialite specialite, String description) {
        DemandeRDV d = new DemandeRDV();
        d.setPatient(patient);
        d.getPatient().setId(patient.getId());
        d.setSpecialite(specialite);
        d.setDescription(description);
        d.setCreatedAt(LocalDateTime.now());
        d.setStatut(Statut.EN_COURS);
        return d;
    }

    @Override
    public void addDemand(DemandeRDV demande) {
        if (demande.getCreatedAt() == null) demande.setCreatedAt(LocalDateTime.now());
        if (demande.getStatut() == null) demande.setStatut(Statut.EN_COURS);
        demandeRepo.insertDemande(demande);
    }

    @Override
    public List<DemandeRDV> searchDemand(Patient patient) {
        return demandeRepo.selectDemande(patient.getId());
    }

    @Override
    public List<DemandeRDV> searchApointment(int patientId) {
        return demandeRepo.selectAppointment(patientId);
    }

    @Override
    public List<DemandeRDV> filterDemandByStatus(String statut) {
        return demandeRepo.selectDemandeByStatut(statut);
    }


    @Override
    public void changeDemandeStatut(DemandeRDV demande, Statut statut) {
        if (demande != null && demande.getId() != null) {
            demandeRepo.updateStatut(demande.getId(), statut);
            demande.setStatut(statut);
        }
    }

    @Override
    public List<DemandeRDV> getAllDemande() {
        return demandeRepo.findAll();
    }

    @Override
    public DemandeRDV serchDemandeById(int demandeId) {
        return demandeRepo.findById(demandeId);
    }
}

File: src/App/service/DemandeRDVService.php
<?php

namespace App\service;

use App\Core\Database;
use App\entity\DemandeRDV;
use App\entity\Patient;
use App\entity\Specialite;
use App\entity\Statut;
use App\repository\DemandeRDVRepository;
use App\service\interface\DemandeRDVServiceImpl;
use DateTime;
use PDO;

class DemandeRDVService implements DemandeRDVServiceImpl
{

    private static ?DemandeRDVService $instance = null;

    private DemandeRDVRepository $demandeRepo;

    private function __construct()
    {
        $this->demandeRepo = DemandeRDVRepository::getInstance();
    }


    public static function getInstance(): DemandeRDVService
    {
        if (self::$instance === null) {
            self::$instance = new DemandeRDVService();
        }
        return self::$instance;
    }


    public function creatDemand(Patient $patient, Specialite $specialite, string $description): DemandeRDV
    {
        $d = new DemandeRDV();
        $d->setPatient($patient);
        $d->getPatient()->setId($patient->getId());
        $d->setSpecialite($specialite);
        $d->setDescriptiion($description);
        $d->setCreateAt(new DateTime());
        $d->setStatut(Statut::EN_COURS);

        return $d;
    }

    public function addDemand(DemandeRDV $demande): void
    {            
        if ($demande->getCreateAt() == null) {
           $demande->setCreateAt(new DateTime());
        }

        if ($demande->getStatut() == null) {
            $demande->setStatut(Statut::EN_COURS);
         }

         $this->demandeRepo->insertDemande($demande);;
    }

    public function searchDemand(Patient $patient): array
    {
        return $this->demandeRepo->selectDemande($patient->getId());
    }

    public function searchApointment(Patient $patient): array
    {
        return $this->demandeRepo->selectApointment($patient->getId());
    }

    public function filterDemandByStatus(Patient $patient, string $statut): array
    {
        return $this->demandeRepo->selectDemandeByStaut(patientId:$patient->getId(), statut:$statut);
    }
    
    public function searchDemandeById(int $id): DemandeRDV
    {
        return $this->demandeRepo->findById($id);
    }

    public function getAllDemandes(): array
    {
        return $this->demandeRepo->findAll();
    }
    
    public function changedeamndeStatus(DemandeRDV $demande, Statut $statut): void
    {
        $this->demandeRepo->updateStatus(demandeId:$demande->getId(), statut:$statut);
    }
}